components/Collections/tests/unit/CollectionsStylerTests.m

```diff
@@ -0,0 +1,136 @@
+}
+static MDCCollectionViewLayoutAttributes* footer0() {
+  return [MDCCollectionViewLayoutAttributes
+      layoutAttributesForSupplementaryViewOfKind:UICollectionElementKindSectionFooter
+                                   withIndexPath:[NSIndexPath indexPathForItem:0 inSection:0]];
+}
+static MDCCollectionViewLayoutAttributes* footer1() {
+  return [MDCCollectionViewLayoutAttributes
+      layoutAttributesForSupplementaryViewOfKind:UICollectionElementKindSectionFooter
+                                   withIndexPath:[NSIndexPath indexPathForItem:1 inSection:1]];
+}
+
+// Implementation of a MDCCollectionViewStylingDelegate for testing purpose.
+@interface CollectionsStylerTestsDelegate : NSObject <MDCCollectionViewStylingDelegate>
+
+@end
+
+@implementation CollectionsStylerTestsDelegate
+
+- (BOOL)collectionView:(UICollectionView*)collectionView
+    shouldHideItemSeparatorAtIndexPath:(NSIndexPath*)indexPath {
+  return indexPath.section == 0 && indexPath.item == 1;
+}
+
+- (BOOL)collectionView:(UICollectionView*)collectionView
+    shouldHideHeaderSeparatorForSection:(NSInteger)section {
+  return section == 1;
+}
+
+- (BOOL)collectionView:(UICollectionView*)collectionView
+    shouldHideFooterSeparatorForSection:(NSInteger)section {
+  return section == 0;
+}
+
+@end
+
+@interface CollectionsStylerTests : XCTestCase
+@end
+
+@implementation CollectionsStylerTests
+
+- (void)testHideSeparatorWithDelegate {
+  // Given
+  UICollectionView* collectionView =
+      [[UICollectionView alloc] initWithFrame:CGRectZero
+                         collectionViewLayout:[[UICollectionViewFlowLayout alloc] init]];
+  CollectionsStylerTestsDelegate* delegate = [[CollectionsStylerTestsDelegate alloc] init];
+  MDCCollectionViewStyler* styler =
+      [[MDCCollectionViewStyler alloc] initWithCollectionView:collectionView];
+  styler.delegate = delegate;
+
+  // Then
+  XCTAssertFalse([styler shouldHideSeparatorForCellLayoutAttributes:cell00()]);
+  XCTAssertTrue([styler shouldHideSeparatorForCellLayoutAttributes:cell01()]);
+  XCTAssertFalse([styler shouldHideSeparatorForCellLayoutAttributes:header0()]);
+  XCTAssertTrue([styler shouldHideSeparatorForCellLayoutAttributes:header1()]);
+  XCTAssertTrue([styler shouldHideSeparatorForCellLayoutAttributes:footer0()]);
+  XCTAssertFalse([styler shouldHideSeparatorForCellLayoutAttributes:footer1()]);
+}
+
+- (void)testHideSeparatorWithoutDelegate {
+  // Given
+  UICollectionView* collectionView =
+      [[UICollectionView alloc] initWithFrame:CGRectZero
+                         collectionViewLayout:[[UICollectionViewFlowLayout alloc] init]];
+  MDCCollectionViewStyler* styler =
+      [[MDCCollectionViewStyler alloc] initWithCollectionView:collectionView];
+  styler.delegate = nil;
+
+  // When
+  styler.shouldHideSeparators = YES;
+
+  // Then
+  XCTAssertTrue([styler shouldHideSeparatorForCellLayoutAttributes:cell01()]);
```

Comment:
Please verify all 6 cases in each test, I think it makes it clearer that all cells are configured by the property.